### ISSUE TITLE
2FA doesn't work when Security Key is enabled

### ISSUE THREAD

**Issue by maxromanovsky** [icloud-photos-downloader/icloud_photos_downloader#583]
## Overview
[TIP]:  # ( DO NOT include your login-information when pasting in program output! )
[NOTE]: # ( Give a BRIEF summary about your problem )

2FA doesn't work when Security Key is enabled

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Enable FIDO key as 2FA on iCloud account
2. Run `icloudpd --recent=1 -u example@gmail.com -d tmp`
3. Choose SMS or TOTP as 2FA 

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )
2FA works with the FIDO key or there is an error message that FIDO is unsupported

## Actual Behavior
[NOTE]: # ( Tell us what actually happens )
- Error if SMS is selected as 2FA
- App hangs if TOTP is selected as 2FA

SMS = Error:
```shell
2023-02-12 10:42:22 DEBUG    Authenticating...
iCloud Password:
2023-02-12 10:42:27 INFO     Two-step/two-factor authentication is required!
  0: SMS to *******XX
  1: Enter two-factor authentication code
Please choose an option: [0]: 0
2023-02-12 10:42:30 ERROR    Failed to send two-factor authentication code
```

TOTP = hangs forever:
```shell
iCloud Password:
2023-02-12 10:37:08 INFO     Two-step/two-factor authentication is required!
  0: SMS to *******XX
  1: Enter two-factor authentication code
Please choose an option: [0]: 1
Please enter two-factor authentication code: ^CAborted!
```

## Context
[NOTE]: # ( Give us any additional information you may have. )

**Comment by ChiefGyk3D** [comment]
I've been able to replicate this issue as well in my docker running on my NAS

**Comment by AndreyNikiforov** [comment]
@maxromanovsky @ChiefGyk3D can you try entering code generated by Security Key Device at the "Please enter two-factor authentication code" prompt, please?

**Comment by lawrence-forooghian** [comment]
@AndreyNikiforov What code do you mean? Are you thinking of a TOTP code, such as that generated by an app like Google Authenticator? That’s not how FIDO keys work; the application communicates directly with the security key (over USB or Bluetooth or NFC) and there is no code exposed to the user.

**Comment by AndreyNikiforov** [comment]
> @AndreyNikiforov What code do you mean? Are you thinking of a TOTP code, such as that generated by an app like Google Authenticator? That’s not how FIDO keys work; the application communicates directly with the security key (over USB or Bluetooth or NFC) and there is no code exposed to the user.

My experience is limited to Yubi Leave-in key (not linked to iCloud). If I have an entry field in focus and touch the key, it spits characters (as if it is a keyboard). If the same behavior is true for other keys that are used (no matter of communication technology), then it should be trivial to test if that stream of generated characters from the device registered with iCloud works out of the box. That was the test I asked for, sorry for confusion.

**Comment by lawrence-forooghian** [comment]
(Unlike Yubico OTP, FIDO requires _two-way_ communication between the computer and security key, which means that the fix for this will be more involved than simply getting some input from the user.)

**Comment by ChiefGyk3D** [comment]
> @maxromanovsky @ChiefGyk3D can you try entering code generated by Security Key Device at the "Please enter two-factor authentication code" prompt, please?

There is no "code" to enter when it comes to the FIDO standard. That's what makes it more secure, it has to communicate with our USB keys in a two way fashion.

FIDO is becoming standardized across the industry Apple, Microsoft, and many other companies are using it. There are libraries to pass it through in Linux and more. I am even using a FIDO login to get into this GitHub for years. https://fidoalliance.org/how-fido-works/

**Comment by iain-henderson** [comment]
It does seem that Yubico provides a Python Library
https://github.com/Yubico/python-fido2

**Comment by eliliam** [comment]
I am having the same issue, has there been any development into resolving this? If not, is there a way to use SMS to authenticate? It appears that option 0(SMS code) always errors out when you have configured FIDO security keys.

**Comment by AndreyNikiforov** [comment]
> I am having the same issue, has there been any development into resolving this? If not, is there a way to use SMS to authenticate? It appears that option 0(SMS code) always errors out when you have configured FIDO security keys.

Can you use SMS for web login to icloud.com? `icloudpd` just simulates web access, so if SMS is blocked for web access once you add fido, then I don't see a way to have `icloudpd` working for that account.

**Comment by gmemstr** [comment]
Yeah, it does block SMS, or any other kind of 2FA, for web access. Apple's security key implementation is also pretty half-baked, and will complain if you're using anything other than Safari or Chrome.

**Comment by AndreyNikiforov** [comment]
FIDO is [not supported](https://github.com/icloud-photos-downloader/icloud_photos_downloader/blob/master/FAQ.md#fido). changing `bug` to `enhancement`

**Comment by MarcHagen** [comment]
fyi, for those who are still wanting a solution.
This is not the most efficient or elegant way to do, but seems to work for me.

And only for people that are handy with Browser developer tools
**If you are unsure about this or are not familiar with JSON, manual file editing or searching for cookie information in developer tools, you should skip this section.**

<details>
<summary>You have been warned, no further support provided</summary>

Open your browser at icloud.com and open up the development tools.

Navigate to the application tab and then cookies.

Go through and tick `Keep me signed in` and login with your email.

SELECT PASSWORD LOGIN (in my case i also have passkey)

Login with your security key

Trust the browser

Now back to the application tab and you should see `icloud.com` and `idmsa.apple.com`

for icloudpd search your keyring file (your username without any chars)
This file should contain `Set-Cookie3` with `X-APPLE-` lines.

Update the values from `idmsa.apple.com` cookies to the file
- acn01
- aasp
- aa

Update the values from `icloud.com` cookies to the file
- X-APPLE-UNIQUE-CLIENT-ID
- X-APPLE-WEBAUTH-LOGIN
- X-APPLE-WEBAUTH-VALIDATE
- X-APPLE-WEBAUTH-USER
- X_APPLE_WEB_KB-(with some random key)
- X-APPLE-DS-WEB-SESSION-TOKEN

Now note how the `"` are `\"`

And that should be it, until the refresh key stuff on the apple side expires and you need to do it all over again. Happy days!

_no support for me, again you should be knowing your way around this_ 

</details>

**Comment by OmarWazzan** [comment]
Any chance FIDO support would be added?

I have been using this for a while, and recently added security keys to my icloud account, which now results in this amazing tool no longer working :(

**Comment by AndreyNikiforov** [comment]
> Any chance FIDO support would be added?
> 

 AFAIK nobody committed to reverse engineering FIDO for iCloud

**Comment by bbenich** [comment]
> > Any chance FIDO support would be added?
> 
> AFAIK nobody committed to reverse engineering FIDO for iCloud

I'm taking a stab at it, but no promises. I want to start using this tool but not interested in removing security keys from my account.

**Comment by bbenich** [comment]
I'm successfully pulling my list of tokens and getting a key challenge from the Apple server, so I guess I'm closer than I thought to having a working prototype. The rough Schema I'm looking at: `{
    "fsaChallenge": {
      "challenge": "...",
      "keyHandles": ["...", "..."],
      "rpId": "apple.com",
      "allowedCredentials": "...,..."
    },
    "keyNames": ["Yubi-Key-A", "Yubi-Key-B"]
  }`

**Comment by bbenich** [comment]
> Any chance FIDO support would be added?
> 
> I have been using this for a while, and recently added security keys to my icloud account, which now results in this amazing tool no longer working :(
Thanks to [timlaing/pyicloud](https://github.com/timlaing/pyicloud). I had to temporarily spoof `apple.com` in `/etc/hosts` and use a self-signed certificate to get the security token to accept the challenge, but it's working.
<img width="1491" height="1184" alt="Image" src="https://github.com/user-attachments/assets/4a0d5aa2-3c42-4016-b9f3-1733cbd7d1bc" />
<img width="1505" height="1195" alt="Image" src="https://github.com/user-attachments/assets/5d4194b8-f031-4a99-ade5-56397b97668f" />
<img width="1695" height="1164" alt="Image" src="https://github.com/user-attachments/assets/1090a7b7-915e-4402-b608-f9ae2a039980" />

**Comment by bbenich** [comment]
> explain what is needed for FIDO to work

I had started building out FIDO support from scratch, but later found [timlaing/pyicloud](https://github.com/timlaing/pyicloud) and used their implementation to inform mine. It seems like a better solution to trash my implementation and re-implement with `timlaing/pyicloud` as a dependency. However, if we can't get it working without spoofing `apple.com`, it might be better to implement a separate authentication app that will export cookies that you can then import into `icloudpd`. I do plan on doing all this work and seeing that a working implementation gets merged in, as my personal time permits. In the meantime, I had AI summarize my work and pushed it to my fork [bbenich/icloudpd-fido](https://github.com/bbenich/icloudpd-fido).

**Comment by Alexivia** [comment]
@bbenich, are you planning on merging your implementation into the main repo? Sounds promising, although I have not tested it yet myself...
